bedrock/entity/workflows.py

```diff
@@ -0,0 +1,169 @@
+                    self.prompt_with_history(prompt),
+                    schedule_to_close_timeout=timedelta(seconds=20),
+                )
+
+                workflow.logger.info(f"{response}")
+
+                # Append the response to the conversation history
+                self.conversation_history.append(("response", response))
+
+                # Continue as new every x conversational turns to avoid event
+                # history size getting too large. This is also to avoid the
+                # prompt (with conversational history) getting too large for
+                # AWS Bedrock.
+
+                # We summarize the chat to date and use that as input to the
+                # new workflow
+                if len(self.conversation_history) >= self.continue_as_new_per_turns:
+                    # Summarize the conversation to date using Amazon Bedrock
+                    self.conversation_summary = await workflow.start_activity_method(
+                        BedrockActivities.prompt_bedrock,
+                        self.prompt_summary_from_history(),
+                        schedule_to_close_timeout=timedelta(seconds=20),
+                    )
+
+                    workflow.logger.info(
+                        "Continuing as new due to %i conversational turns."
+                        % self.continue_as_new_per_turns,
+                    )
+
+                    workflow.continue_as_new(
+                        args=[
+                            BedrockParams(
+                                self.conversation_summary,
+                                self.prompt_queue,
+                            )
+                        ]
+                    )
+
+                continue
+
+            # If end chat signal was sent
+            if self.chat_ended:
+                # The workflow might be continued as new without any
+                # chat to summarize, so only call Bedrock if there
+                # is more than the previous summary in the history.
+                if len(self.conversation_history) > 1:
+                    # Summarize the conversation to date using Amazon Bedrock
+                    self.conversation_summary = await workflow.start_activity_method(
+                        BedrockActivities.prompt_bedrock,
+                        self.prompt_summary_from_history(),
+                        schedule_to_close_timeout=timedelta(seconds=20),
+                    )
+
+                workflow.logger.info(
+                    "Chat ended. Conversation summary:\n"
+                    + f"{self.conversation_summary}"
+                )
+
+                return f"{self.conversation_history}"
+
+    @workflow.signal
+    async def user_prompt(self, prompt: str) -> None:
+        # Chat ended but the workflow is waiting for a chat summary to be generated
+        if self.chat_ended:
+            workflow.logger.warn(f"Message dropped due to chat closed: {prompt}")
+            return
```

Comment:
This is an ok approach, though often something like prompt + end will come in at what appears to be "same time" from separate clients, but server happens to get end chat first which means data loss. But that's probably ok for this sample. If workflow update were stable we could return back an error instead of warn.

Reply:
Agree, but I also think that for this example it's okay. For production we'd want to be more careful about how we handle this case, but the tradeoff is additional code/complexity which is probably too much for this sample.